edb/pgsql/compiler/relgen.py

```diff
@@ -2386,32 +2406,77 @@ def process_set_as_agg_expr(
     if expr.func_initial_value is not None:
         iv_ir = expr.func_initial_value.expr
 
-        if newctx.expr_exposed and serialization_safe:
+        if serialization_safe and aspect == 'serialized':
             # Serialization has changed the output type.
             with newctx.new() as ivctx:
-                ivctx.expr_exposed = True
                 iv = dispatch.compile(iv_ir, ctx=ivctx)
+
                 iv = output.serialize_expr_if_needed(
                     iv, path_id=ir_set.path_id, ctx=ctx)
                 set_expr = output.serialize_expr_if_needed(
                     set_expr, path_id=ir_set.path_id, ctx=ctx)
         else:
-            iv = dispatch.compile(iv_ir, ctx=newctx)
+            with newctx.new() as ivctx:
+                iv = dispatch.compile(iv_ir, ctx=ivctx)
+
+        pathctx.put_path_var(
+            stmt, ir_set.path_id, set_expr, aspect=aspect, env=ctx.env)
+        pathctx.get_path_output(
+            stmt, ir_set.path_id, aspect=aspect, env=ctx.env)
+
+        assert wrapper
+        set_expr = pgast.CoalesceExpr(
+            args=[stmt, iv], ser_safe=serialization_safe)
+
+        pathctx.put_path_var(
+            wrapper, ir_set.path_id, set_expr, aspect=aspect, env=ctx.env)
+        stmt = wrapper
+
+    pathctx.put_path_var_if_not_exists(
+        stmt, ir_set.path_id, set_expr, aspect=aspect, env=ctx.env)
+
+    return new_stmt_set_rvar(ir_set, stmt, ctx=ctx)
 
-        pathctx.put_path_value_var(stmt, ir_set.path_id, set_expr, env=ctx.env)
-        pathctx.get_path_value_output(stmt, ir_set.path_id, env=ctx.env)
 
+def process_set_as_agg_expr(
+        ir_set: irast.Set, stmt: pgast.SelectStmt, *,
+        ctx: context.CompilerContextLevel) -> SetRVars:
+    expr = ir_set.expr
+    assert isinstance(expr, irast.FunctionCall)
+
+    # If the func has an initial val, we need to create a wrapper to put
+    # the coalesces we insert in
+    wrapper = None
+    if expr.func_initial_value is not None:
         with ctx.subrel() as subctx:
             wrapper = subctx.rel
-            set_expr = pgast.CoalesceExpr(
-                args=[stmt, iv], ser_safe=serialization_safe)
 
-            pathctx.put_path_value_var(
-                wrapper, ir_set.path_id, set_expr, env=ctx.env)
-            stmt = wrapper
+    # When in a serialization context, we need to compute both a value
+    # version and a serialized version of the aggregate function call,
+    # since we may need both (Consider `WITH X := array_agg(User), ...`).
 
-    pathctx.put_path_value_var_if_not_exists(
-        stmt, ir_set.path_id, set_expr, env=ctx.env)
+    cctx = ctx.subrel() if wrapper else ctx.new()
+    with cctx as xctx:
+        xctx.expr_exposed = False
+        process_set_as_agg_expr_inner(
+            ir_set, xctx.rel, aspect='value', wrapper=wrapper,
+            ctx=xctx)
+
+    # Though if the result type is a scalar, the value should be good
+    # enough, so don't generate bad code.
```

Comment:
Bad code?  I guess `compile_output` will force scalar array serialization anyway (I'd mention that in the comment), but I wonder why this formulation needs to exclude scalars specifically.

Reply:
By bad code I just mean "suboptimal code"; if it's just scalars, then we're fine not providing a "serialized" output, which means we can skip compiling it.